File: packages/server/apps/ai-agent-worker/src/support/llm/pricing.ts
import type { AgentQueryUsage } from "./agent.usage.js";

// 백만 토큰당 USD이며 이름 부분일치로 버전 접미사가 붙은 모델도 해석한다.
interface ModelRate {
    readonly input: number;
    readonly output: number;
    readonly cacheWrite: number;
    readonly cacheRead: number;
}

const MODEL_RATES: ReadonlyArray<readonly [string, ModelRate]> = [
    ["sonnet", { input: 3.0, output: 15.0, cacheWrite: 3.75, cacheRead: 0.3 }],
    ["haiku", { input: 1.0, output: 5.0, cacheWrite: 1.25, cacheRead: 0.1 }],
];

function rateFor(model: string): ModelRate | null {
    const lowered = model.toLowerCase();
    for (const [key, rate] of MODEL_RATES) {
        if (lowered.includes(key)) return rate;
    }
    return null;
}

/** 단가를 모르는 모델은 예산을 집행할 수 없으므로 집행 가능 여부를 먼저 묻는다. */
export function canEstimateCost(model: string): boolean {
    return rateFor(model) !== null;
}

/** 모델이나 사용량을 모르면 오도하지 않도록 null을 낸다. */
export function estimateCostUsd(model: string, usage: AgentQueryUsage | null): number | null {
    const rate = rateFor(model);
    if (rate === null || usage === null) return null;
    const cost =
        (usage.inputTokens * rate.input
            + usage.outputTokens * rate.output
            + usage.cacheCreationTokens * rate.cacheWrite
            + usage.cacheReadTokens * rate.cacheRead)
        / 1_000_000;
    return Math.round(cost * 1_000_000) / 1_000_000;
}

File: packages/agents/src/agent_graph/agents/runtime/pricing.py
"""Python-native 그래프의 지출 서킷브레이커용 근사 단가표."""

from __future__ import annotations

from dataclasses import dataclass

from ..shared.models import UsageDTO


@dataclass(frozen=True)
class ModelRate:
    """모델별 백만 토큰당 달러 단가다."""

    input: float
    output: float
    cache_write: float
    cache_read: float


_RATES: dict[str, ModelRate] = {
    "sonnet": ModelRate(input=3.0, output=15.0, cache_write=3.75, cache_read=0.30),
    "haiku": ModelRate(input=1.0, output=5.0, cache_write=1.25, cache_read=0.10),
}


def _rate_for(model: str) -> ModelRate | None:
    lowered = model.lower()
    for key, rate in _RATES.items():
        if key in lowered:
            return rate
    return None


# 이 값은 그래프 내부 예산 상한에만 쓰고 보고·저장용 costUsd는 ai-agent-worker가 별도 단가표로 환산한다.
def estimate_cost_usd(model: str, usage: UsageDTO | None) -> float | None:
    rate = _rate_for(model)
    if rate is None or usage is None:
        return None
    cost = (
        usage.inputTokens * rate.input
        + usage.outputTokens * rate.output
        + usage.cacheCreationTokens * rate.cache_write
        + usage.cacheReadTokens * rate.cache_read
    ) / 1_000_000
    return round(cost, 6)
